### PR TITLE
feat: Enable remote pilot logging using REST interface to Tornado.

### PR DESCRIPTION
**Pull request by martynia**
This is a revived remote pilot logger using REST interface to Tornado. No other interfaces (local file, message queue) were modified.

Superseeds #152  (targeting devel)

### CODE REVIEW

File: Pilot/PilotLogger.py [hunk 1]
```diff
@@ -217,7 +254,12 @@ def singleWord(arg):
...
-    logger.sendMessage(messageContent=" ".join(args.message), source=args.source, phase=args.phase, status=args.status)
+    logger.sendMessage(
+        messageContent=" ".join(args.message),
+        source=args.source,
+        phase=args.phase,
+        status=args.status,
+    )
 
 
 if __name__ == "__main__":
```
[the remark this answers is not in the document]
Comment (martynia) [reply Pilot/PilotLogger.py:265]: I don't follow. Are you referring to the main() method ? This is one of those things which never runs (at least in the current system)

File: Pilot/PilotLogger.py [hunk 2]
```diff
@@ -122,41 +123,71 @@ def __init__(
...
-        logging.debug("In sendMessage of PilotLogger")
+
         if not self._isCorrectStatus(status):
-            logging.error("status: " + str(status) + " is not correct")
+            self.logger.error("status: %s is not correct", str(status))
             return False
         myUUID = getPilotUUIDFromFile(self.params["FileWithID"])
-        message = generateDict(myUUID, generateTimeStamp(), source, phase, status, messageContent)
+        message = generateDict(
+            myUUID, generateTimeStamp(), source, phase, status, messageContent
```
[the remark this answers is not in the document]
Comment (martynia) [reply Pilot/PilotLogger.py:171]: OK, I'll look into this.

File: Pilot/PilotLogger.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (martynia) [reply Pilot/PilotLogger.py:265]: To clarify, currently, the classic logger (always on) prints to stdout which is captured. The extended logger logs by other means on top of that.

[the remark this answers is not in the document]
Comment (chrisburr) [reply Pilot/PilotLogger.py:265]: (the line this is referring isn't related to the comment, just a way of being able to make a thread)

I mean to allow us to capture things which aren't explicitly logged from within Python. For example segfaults, kill signals and bus errors.

File: Pilot/PilotLogger.py [hunk 3]
```diff
@@ -122,41 +123,71 @@ def __init__(
...
+        )
         if not isMessageFormatCorrect(message):
-            logging.warning("Message format is not correct.")
+            self.logger.warning("Message format is not correct.")
             return False
+        self.logger.debug("sendMessage: %s ", message)
         encodedMsg = encodeMessage(message)
         return self.messageSender.sendMessage(encodedMsg, flag=status)
```
[the remark this answers is not in the document]
Comment (martynia) [reply Pilot/PilotLogger.py:178]: Would you like to accumulate all the messages for one pilot and then send them ? The cache would have to saved in the case of an exit(). This the classic resource clear-up problem. In the current file cache this is the logFinaliser wrapper which does that. The smaller the batches, the safer the whole system is, even if the finaliser fails for some reason. We don't want to lose the whole record, do we ?

[the remark this answers is not in the document]
Comment (chrisburr) [reply Pilot/PilotLogger.py:178]: It does't need to be sent in a single request at the end, but reducing the load would be helpful. Naively I think limiting to something like 1 request every 10 seconds is a reasonable trade off. Batching reqests would be probably be easier to implement with https://github.com/DIRACGrid/Pilot/pull/153#discussion_r829161740.

For reference in LHCb LHCb, we have something like 2000 lines per log from 250,000 pilots per day.[1] That means this needs to be able to handle at least 6000 log entries per second on average, likely with spikes which are an order of magniude higher. If there are operational issues the number of pilots will also spike.

[1] This is an under-estimate

File: Pilot/PilotLogger.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (martynia) [reply Pilot/PilotLogger.py:265]: Do you mean wrapping the pilot executable ? The python script created by `src/DIRAC/WorkloadManagementSystem/Utilities/PilotWrapper.py` ? Should this new wrapper fork off the `stdout/stderr` and send them to a remote service ? Is it what you are suggesting ? The "normal" logger would behave as it does now, right ?

[the remark this answers is not in the document]
Comment (chrisburr) [reply Pilot/PilotLogger.py:265]: That would be my suggestion for how to implement this, though @fstagni should also give input for if he agrees.

The normal logger could even be adapted to print a JSON per line to make it easier to ingest.

[the remark this answers is not in the document]
Comment (chrisburr) [reply Pilot/PilotLogger.py:265]: A wrapper feels far simplier to me that trying to hook into the logging machinery. Especially if batching HTTP requests is needed for performance.

> TBH I don't think it's up to DIRAC to notice about these messages. 

Why? And who should notice? Segfaults are likely a DIRAC(OS) issue and the others make it easier to ticket sites. It also alows use to capture issues with getting the pilot files.

[the remark this answers is not in the document]
Comment (fstagni) [reply Pilot/PilotLogger.py:265]: You might well be right on the point of catching segfaults and so on. 

Anyway, the PilotWrapper that is created on DIRAC side has a downside: it's a SiteDirector creation at runtime. Meaning that, if there are sites/computing elements to which the pilots are NOT sent by SiteDirectors (e.g. clouds, but also in LHCb the HLT farm -- yes these use another "pilot wrapper"), then those won't use such PilotWrapper. And often these sites are those for which we have the least monitoring. So, I think that Pilot logging needs to come from within these modules.

File: Pilot/pilotCommands.py [hunk 4]
```diff
@@ -49,18 +49,59 @@ def __init__(self, pilotParams):
...
+            return func(self)
+        except SystemExit:  # or Exception ?
+            if not self.log.isPilotLoggerOn:
+                self.log.info("Remote logger is not active, nothing to do")
+            else:
+                try:
+                    myUUID = getPilotUUIDFromFile(
+                        self.log.pilotLogger.params["FileWithID"]
+                    )
+                    self.log.info("Finalising the remote logger for pilot: %s" % myUUID)
```
[the remark this answers is not in the document]
Comment (martynia) [reply Pilot/pilotCommands.py:84]: This is a pilot logger, so your suggestion does not work.

[the remark this answers is not in the document]
Comment (chrisburr) [reply Pilot/pilotCommands.py:84]: Isn't this Python's standard library `logging` module?

```
In [1]: import logging
   ...: log = logging.getLogger("something")
   ...: log.setLevel(logging.INFO)
   ...: log.warning("Hello world %s", 1234)
Hello world 1234
```

File: Pilot/pilotCommands.py [hunk 5]
```diff
@@ -49,18 +49,59 @@ def __init__(self, pilotParams):
...
+    from Pilot.PilotLogger import getPilotUUIDFromFile
 except ImportError:
     from pilotTools import CommandBase, retrieveUrlTimeout
+    from PilotLogger import getPilotUUIDFromFile
 ############################
 
 
+def logFinalizer(func):
+    """
+    PilotCammand decorator. It marks a log file as final so no more messages should be written to it .
```
Comment (aldbr) [inline Pilot/pilotCommands.py:61]:
```suggestion
    PilotCommand decorator. It marks a log file as final so no more messages should be written to it .
```

File: Pilot/PilotLogger.py [hunk 6]
```diff
@@ -108,7 +108,8 @@ def __init__(
           localOutputFile(str): Name of the file that can be used to store the log messages locally.
           fileWithUUID(str): Name of the file used to store the Pilot identifier.
         """
-        logging.debug("In init of PilotLogger")
+        self.logger = logging.getLogger(self.__class__.__name__)
+        self.logger.debug("In init of PilotLogger")
         self.STATUSES = PilotLogger.STATUSES
```
Comment (aldbr) [inline Pilot/PilotLogger.py:113]: I guess this is not needed, is it?

```suggestion
```

Example:
```python
class A():
    b = 5
    def __init__(self):
        self.c = 4

A.b
> 5
A.c
> error
A().b
> 5
A().c
> 4
```

**Review by aldbr — COMMENTED** [review]
It seems like `MessageSender` contains classes that look like `logging.handlers`.
Is there a reason not to use `handlers` to emit messages?

Examples:
- `FileHandler` -> `LocalFileSender` (https://docs.python.org/3/library/logging.handlers.html)
- Custom handlers from developers for REST (https://pypi.org/project/restapi-logging-handler/ nothing since 2018, dead project?), MQ (https://pypi.org/project/python-logging-rabbitmq/ seems specific to `RabbitMQ`)

I am sorry, I have just taken a quick look at the code, I am sure there are many complex components that prevent such an "easy" implementation, I just wanted to know.

Thanks

**Comment by martynia** [comment]
Outdated

**Pull request closed without merging**